### PR TITLE
fix(ms365): validate Retry-After while preserving upstream cooldown floor

### PR DESCRIPTION
**Pull request by kaijakagi-sys**
## Original residual after #13534

Related to #13176. #13534 already fixed the short cooldown bug and is preserved/credited. This update narrows #13299 to our previously submitted validated Retry-After parser and its regression coverage: main still raises ValueError for malformed headers and accepts integers too large for asyncio timers. Both JSON and raw file handlers now use the original shared parser, retaining #13534's max(server delay, min(exponential backoff, 30)) policy.

The zero-header expected sleep is now 2s, not 0s, to preserve the independently merged #13534 contract (https://github.com/BasedHardware/omi/pull/13534, commit19627d93902dd22e7f599a6a1dfb922e7ba938f7). Added the 1s case against the same floor. Other original cases remain intact. This is not a duplicate claim for the cooldown fix.

## Verification

After preserving #13534's zero/one-second floor policy in the handler tests, main342d59fc still fails with 4 assertions and 8 ValueErrors for malformed/oversized headers. The repaired handlers pass. For provenance, unchanged original handlers fail before #13534 (18 assertions/8 errors), pass on our original cc2b4fb5 head, and fail on current main (8 assertions/8 errors, including four expected floor-policy differences).

Updated production-handler suite: 5 tests / 50 subTest cases pass with HTTP/token seams; no credentials/network. Real asyncio timer test immediately cancels scheduled tasks. Live Microsoft Graph/account workflow not exercised.

Both `OMI_PR_BODY_FILE=../PR_BODY.md make preflight` and explicit `pr_preflight.py --lane local --base origin/main --pr-body-file ../PR_BODY.md` pass all 14 checks. Local source uses inherited shallow boundaries; nested-alternate warnings remain, so full repository history is not claimed.

Preserves all fresh-main manifest checks and production changes. Existing three-attempt cap and final GraphError preserved. This is bounded follow-through on an existing discretionary-reward contribution; no approved amount or payment claimed.

Failure-Class: none

### CODE REVIEW

File: plugins/omi-ms365-app/services/graph_client.py [hunk 1]
```diff
@@ -15,6 +15,18 @@
 MAX_RETRIES = 3
 
 
+def _retry_delay(retry_after: str | None, attempt: int) -> int:
+    """Honor Graph's nonnegative delay-seconds; otherwise back off locally."""
+    if retry_after is not None:
+        value = retry_after.strip()
+        if value.isascii() and value.isdecimal():
+            try:
+                return int(value)
```
[the remark this answers is not in the document]
Comment (kaijakagi-sys) [reply plugins/omi-ms365-app/services/graph_client.py]: Confirmed on `2f89afaa`: a 400-digit numeric header reached the real `asyncio.sleep` timer arithmetic and raised `OverflowError`. The shared parser now checks the integer’s float conversion and uses the existing 2/3-second fallback on overflow, preserving supported integer delays without an arbitrary `10**308` cap.

Added JSON/raw regressions for oversized headers and the maximum finite float integer, plus a real event-loop scheduling test with immediate cancellation (no elapsed timer waits). The extended suite fails before the correction with 4 assertion failures and 2 `OverflowError`s, then passes all 5 tests / 46 cases. README and PR description now state the representability boundary.

**Comment by Git-on-my-level** [comment]
Thanks @kaijakagi-sys — this is a well-executed fix for #13176, and the process on the review finding (reproducing the real event-loop `OverflowError` on the first commit, then bounding it) is exactly right. I verified the change end to end; notes per file:

- `plugins/omi-ms365-app/services/graph_client.py` — the new `_retry_delay` helper replaces both `min(int(Retry-After, "2"), 2 ** attempt + 1)` sites (`_request` and `get_bytes`) with one policy: ASCII decimal delay-seconds are honored in full when they survive the `float()` representability check, and everything else (missing, empty, malformed, negative, fractional, or overflow-large) falls back to 2 then 3 seconds. The 3-attempt cap and final `GraphError` behavior are untouched. A `429 Retry-After: 60` now sleeps the requested 60s instead of exhausting the retry budget mid-cooldown — precisely what the issue asked for.
- `plugins/omi-ms365-app/test_graph_retry.py` — hermetic stdlib-only suite; I ran it from a scratch checkout: 5 tests / 46 subTest cases pass in ~0.05s with no credentials or network. Good coverage of both paths (JSON `get` and raw `get_bytes`), 429/503, honored/fallback headers, exhaustion, and non-retryable errors, plus a real `call_later` float-arithmetic check via immediate task cancellation.
- `.github/checks-manifest.yaml` — the new `ms365-graph-retry-tests` entry follows the documented convention (both lanes, path-scoped trigger, no PR-body dependency) and runs the hermetic suite directly.
- `plugins/omi-ms365-app/README.md` — every claim in the new "Retry regression tests" section matches the code I read (representability boundary, fallback values, preserved attempt cap).

The earlier review's overflow finding is resolved on this head: `float(delay)` now rejects integers the event-loop clock cannot represent, with `OverflowError` caught alongside `ValueError`.

One non-blocking consideration: the representability boundary is ~1.7e308 seconds, so effectively any representable delay is honored in full. Real Graph cooldowns are seconds-to-minutes, but a buggy or hostile intermediary returning e.g. `Retry-After: 1000000000` would stall that one tool call for ~31 years inside `asyncio.sleep` rather than exhausting retries and raising `GraphError` (the pre-change failure mode). A coarse upper sanity bound (on the order of minutes) would keep failure latency bounded; alternatively keeping strict adherence to the advertised delay is a defensible reading of the throttling guidance — flagging it for the merge decision.

Leaving the merge call with a human maintainer.

— Automated maintainer feedback on behalf of the Omi maintainers

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by kodjima33 — APPROVED** [review]
backend/community fix: confirmed on main backoff = min(retry_after, 2**attempt+1) capped the actual wait BELOW the server's Retry-After, ignoring MS Graph throttling guidance and risking repeated 429s; also int() on a malformed header could raise. Fix honors the header when a valid nonnegative int, else falls back to exponential backoff. Issue #13176 real, tests added.

**Review by kodjima33 — APPROVED** [review]
Honors Graph Retry-After cooldowns instead of clamping them against exponential backoff; scoped fix, references issue #13176, tests added, confidence 5/5. CLEAN, not fixed on main.

**Comment by Git-on-my-level** [comment]
Follow-up verification on the new head `2e1355b` (merge of current main into `8304e0f`), since both maintainer approvals predate it:

- `plugins/omi-ms365-app/services/graph_client.py`, `plugins/omi-ms365-app/test_graph_retry.py`, and `plugins/omi-ms365-app/README.md` are byte-identical (sha256) to the head @kodjima33 approved on Sep 10, so the earlier per-file review still applies to the code surface; the only new content in this head is the merge of main.
- `.github/checks-manifest.yaml`: the `ms365-graph-retry-tests` entry survived the main merge cleanly (still registered for both lanes with the path-scoped trigger).
- Re-ran `python3 plugins/omi-ms365-app/test_graph_retry.py` in an isolated no-network container at this head: 5 tests / 46 cases pass in ~0.05s, and the `float()`-representability boundary behaves as the README documents (max-float-integer header honored in full; 400-digit values fall back to 2s/3s).

CI note for maintainers: the red **Dart Analyze + Tests** lane is not from this PR (it touches no Dart). The errors are compile-time typedef mismatches between main's own `app/lib/providers/action_items_provider.dart` (`ActionItemsFetcher = Future<ActionItemsResponse?> Function(...)`, reworked with the recent action-items due-window change) and main's tests `app/test/providers/deletion_consistency_test.dart` and `app/test/widgets/chat_content_blocks_test.dart`, whose fetchers return `Future<GeneratedActionItemsResponse>`. That mismatch rides in with the merged main content and will keep the lane red on any PR that merges current main until those main-side tests are fixed — merging this PR neither fixes nor worsens it.

Leaving the merge timing and re-approval on the merged head with a human maintainer; the PR's own surface is verified.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by kodjima33 — APPROVED** [review]
Re-verified after new commits: MS365 Graph Retry-After backoff fix is real, scoped, tested, not yet on main. CI green. DIRTY/conflicting mergeability — leaving at approve, not merging.

**Comment by kaijakagi-sys** [comment]
The conflict noted in the last review is resolved on current head `70cedb83532e47bd25894271caa6967f7df93956`, which is now mergeable. All nine workflow names are green, including [Desktop Swift CI](https://github.com/BasedHardware/omi/actions/runs/34765349376): release compilation, Swift app tests, launcher tests, and the independent phase-verdict gate passed. The release-mode callback regression step was skipped; I am not counting it as tested.

This revised head preserves the cooldown floor merged in #13534 and retains the malformed/unschedulable Retry-After header handling and regression coverage described in the updated PR body. The earlier approvals predate this revision. Could a maintainer review this remaining scope and make the merge decision?

**Comment by Git-on-my-level** [comment]
Verified on the current head `70cedb83` — the narrowed post-#13534 scope checks out:

- `plugins/omi-ms365-app/services/graph_client.py` — the production delta vs main is exactly the `_retry_delay` helper plus its two call sites (`_request` and `get_bytes`). #13534's `max(server delay, min(2 ** attempt + 1, 30))` floor is preserved, and the three-attempt cap and final `GraphError` are untouched.
- `plugins/omi-ms365-app/test_graph_retry.py` — ran the suite from an isolated no-network container at this head: 5 tests / 50 cases pass in ~0.1s, including the real event-loop `call_later` representability check via immediate task cancellation.
- `.github/checks-manifest.yaml` — diffed against main: the only change is the new `ms365-graph-retry-tests` entry; both lanes, path-scoped trigger, and the documented manifest convention.
- `plugins/omi-ms365-app/README.md` — the "Retry regression tests" section matches the verified behavior (2s/3s floor values, float-representability boundary, fallback semantics, preserved attempt cap).

Residual-bug check against current main (`342d59fc`): main's handlers still crash with `ValueError` on malformed `Retry-After` headers (the bare `int(...)` parse) and hand unschedulable huge delays straight to `asyncio.sleep` — so the residual this PR carries is a real, distinct hardening fix, not a duplicate of #13534. The PR body's floor-policy notes (expected 2s sleep for the zero/one-second cases under the #13534 contract) match what the suite actually asserts.

All three maintainer approvals predate this revised head and the scope narrowing, so what remains is a re-approval / merge-timing decision on the residual. The technical surface is verified; leaving that call with a human maintainer.

— Automated maintainer feedback on behalf of the Omi maintainers

---
*by AI on behalf of David — if you need David's attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Comment by Git-on-my-level** [comment]
State update since the last automated check on this head: the PR conflicts with main again, and the conflict is confined to `.github/checks-manifest.yaml`. Since this branch last merged main (`342d59fc`), main added `linear-exact-issue-lookup-tests` at the same spot where this PR inserts `ms365-graph-retry-tests` (top of the `checks:` list) - purely textual, no semantic resolution needed. Merging main and keeping both entries clears it. Nothing else can conflict: `plugins/omi-ms365-app/services/graph_client.py` is unchanged on main since that merge base, the README section is purely additive, and the test file is new.

Re-verified while checking: the production delta remains exactly the `_retry_delay` helper plus its two call sites in `_request`/`get_bytes` (upstream #13534 floor intact), and the hermetic suite passes (5 tests / 50 cases, no network or credentials). Spot-probing `_retry_delay` edge cases (missing, empty, `abc`, `-3`, `1.5`, non-ASCII digits, 400-digit overflow, `120`, whitespace-padded, zero/one-second floor) matches the documented contract - malformed or unschedulable headers fall back to 2s/3s instead of main's bare `int()` crash.

The three earlier approvals all predate this head, so re-approval is still with a human maintainer - a merge-main refresh to clear the manifest conflict would be a natural moment for that final pass.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by aryanorastar — CHANGES_REQUESTED** [review]
Verified on `70cedb83532e`.

Head is CONFLICTING/DIRTY, so this cannot merge even with kodjima APPROVE. Rebase onto current `main` and keep the diff on `.github/checks-manifest.yaml`, `plugins/omi-ms365-app/README.md`, `plugins/omi-ms365-app/services/graph_client.py`, `plugins/omi-ms365-app/test_graph_retry.py`.

kodjima size-gate APPROVE is not a maintainer merge. I am not merging until this is MERGEABLE and green.

**Comment by aryanorastar** [comment]
Closing. Retry-After parsing already landed in #14250.
